components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ConditionalRequestHandler.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Sonatype Nexus (TM) Open Source Version
+ * Copyright (c) 2008-2015 Sonatype, Inc.
+ * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
+ *
+ * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.view;
+
+import javax.annotation.Nonnull;
+
+import org.sonatype.nexus.repository.Repository;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Predicate;
+import com.google.common.net.HttpHeaders;
+
+import static org.sonatype.nexus.repository.http.HttpConditions.requestPredicate;
+import static org.sonatype.nexus.repository.http.HttpMethods.DELETE;
+import static org.sonatype.nexus.repository.http.HttpMethods.GET;
+import static org.sonatype.nexus.repository.http.HttpMethods.HEAD;
+import static org.sonatype.nexus.repository.http.HttpMethods.POST;
+import static org.sonatype.nexus.repository.http.HttpMethods.PUT;
+import static org.sonatype.nexus.repository.http.HttpStatus.NOT_MODIFIED;
+import static org.sonatype.nexus.repository.http.HttpStatus.PRECONDITION_FAILED;
+
+/**
+ * A format-neutral handler for conditional requests.
+ *
+ * @since 3.0
+ */
+public class ConditionalRequestHandler
+    extends ComponentSupport
+    implements Handler
+{
+  @Nonnull
+  @Override
+  public Response handle(@Nonnull final Context context) throws Exception {
+    final Predicate<Response> requestPredicate = requestPredicate(context.getRequest());
+    if (requestPredicate != null) {
+      final String action = context.getRequest().getAction();
+      switch (action) {
+        case GET:
+        case HEAD: {
+          final Response response = context.proceed();
+          if (response.getStatus().isSuccessful()) {
+            if (!requestPredicate.apply(response)) {
+              return new Response.Builder().status(Status.failure(NOT_MODIFIED, "Not modified")).build();
+            }
+          }
+          return response;
+        }
+
+        case POST:
+        case PUT:
+        case DELETE: {
+          final Response getResponse = performGet(context.getRepository(), context.getRequest());
+          if (getResponse.getStatus().isSuccessful() && !requestPredicate.apply(getResponse)) {
+            return new Response.Builder().status(Status.failure(PRECONDITION_FAILED, "Precondition failed")).build();
```

Comment:
Status message set is actually the "reason phrase" as per RFC http://tools.ietf.org/html/rfc7230#section-3.1.2

And should not be used as clients are advised to ignore it completely.

OTOH, added toString method to HTTP conditions, so doing a toString() on returned non-null `requestPredicate` should describe just nicely what conditions are (see added DEBUG log too)